### PR TITLE
Add metrics representing in-progress state.

### PR DESCRIPTION
**Pull request by kfyharukz**
Signed-off-by: Kazuhito MATSUDA <kazuhito.matuda@gmail.com>

### CODE REVIEW

File: server/backup_binlog.go [hunk 1]
```diff
@@ -76,20 +76,25 @@ func (s *backupBinlogService) FlushAndBackupBinlog(ctx context.Context, req *age
 		Bucket: aws.String(req.BucketName),
 		Key:    aws.String(req.BackupId),
 	})
+
 	if err == nil {
 		s.agent.sem.Release(1)
 		return nil, status.Errorf(codes.InvalidArgument, "the requested backup has already completed: BackupId=%s", req.BackupId)
 	}
+	metrics.SetBinlogBackupInProgressMetrics(s.agent.clusterName, true)
```
Comment (morimoto-cybozu) [inline server/backup_binlog.go:84]: This seems too early.
Placing this just before `flushBinaryLogs()` would be appropriate.

File: server/backup_binlog.go [hunk 2]
```diff
@@ -160,7 +170,11 @@ func (s *backupBinlogService) FlushBinlog(ctx context.Context, req *agentrpc.Flu
 	if !s.agent.sem.TryAcquire(1) {
 		return nil, status.Error(codes.ResourceExhausted, "another request is under processing")
 	}
-	defer s.agent.sem.Release(1)
+	metrics.SetBinlogBackupInProgressMetrics(s.agent.clusterName, true)
+	defer func() {
+		s.agent.sem.Release(1)
+		metrics.SetBinlogBackupInProgressMetrics(s.agent.clusterName, false)
+	}()
```
Comment (morimoto-cybozu) [inline server/backup_binlog.go:177]: Do we need to set the "in-progress" flag just for flushing binlog?

File: server/clone.go [hunk 3]
```diff
@@ -65,9 +65,13 @@ func (s *cloneService) Clone(ctx context.Context, req *agentrpc.CloneRequest) (*
 	if !s.agent.sem.TryAcquire(1) {
 		return nil, status.Error(codes.ResourceExhausted, "another request is under processing")
 	}
+	metrics.SetCloneInProgressMetrics(s.agent.clusterName, true)
```
Comment (morimoto-cybozu) [inline server/clone.go:68]: This also can be postponed.
Just before `clone()`, or inside of `clone()` as other metrics.

File: metrics/updater.go [hunk 4]
```diff
@@ -3,47 +3,53 @@ package metrics
...
 func UpdateCloneDurationSecondsMetrics(clusterName string, durationSeconds float64) {
 	cloneDurationSecondsMetrics.WithLabelValues(clusterName).Observe(durationSeconds)
 }
+func SetCloneInProgressMetrics(clusterName string, inProgress bool) {
+	if inProgress {
+		cloneInProgressMetrics.WithLabelValues(clusterName).Set(1.0)
+		return
+	}
+	cloneInProgressMetrics.WithLabelValues(clusterName).Set(0.0)
+}
```
Comment (morimoto-cybozu) [inline metrics/updater.go:35]: Let's prepare `SetDumpInProgressMetrics()` too, though it will not be used for now.

**Review by morimoto-cybozu — CHANGES_REQUESTED** [review]

**Comment by kfyharukz** [comment]
@morimoto-cybozu The in-progress metrics are used for alerting long-running backup/clone async processes, so I fixed the timing to go up the metrics to that async process start.

File: server/backup_binlog.go [hunk 5]
```diff
@@ -160,7 +166,6 @@ func (s *backupBinlogService) FlushBinlog(ctx context.Context, req *agentrpc.Flu
 	if !s.agent.sem.TryAcquire(1) {
 		return nil, status.Error(codes.ResourceExhausted, "another request is under processing")
 	}
-	defer s.agent.sem.Release(1)
```
Comment (morimoto-cybozu) [inline server/backup_binlog.go:163]: Removed accidentally.

File: server/backup_binlog_test.go [hunk 6]
```diff
@@ -120,14 +120,25 @@ func testBackupBinlog() {
...
 		Eventually(func() error {
 			if agent.sem.TryAcquire(1) {
 				agent.sem.Release(1)
 				return nil
 			}
-			return errors.New("backup process is still working")
+			return errors.New("process is still working")
 		}, 30*time.Second).Should(Succeed())
 
+		By("checking in-progress metric is set")
```
Comment (morimoto-cybozu) [inline server/backup_binlog_test.go:138]: The word "cleared" is used in clone_test.go.

```suggestion
		By("checking in-progress metric is cleared")
```

**Review by morimoto-cybozu — CHANGES_REQUESTED** [review]

File: server/backup_binlog.go (hunk 5 above)
Comment (kfyharukz) [reply server/backup_binlog.go]: 🙇

File: server/backup_binlog_test.go (hunk 6 above)
Comment (kfyharukz) [reply server/backup_binlog_test.go]: 🙇🙇

**Review by morimoto-cybozu — APPROVED** [review]
LGTM

**Review by masa213f — APPROVED** [review]
LGTM

**Pull request merged**